### DISCUSSION TITLE
Spalart-Allmaras 2D Axisymmetric Source Term Question

### DISCUSSION THREAD

**Discussion by mskim0401** [su2code/SU2#1315] [Q&A]
Hi.
I'm trying to run 2D axisymmetric case using RANS SA and SST model.
I ran my case using SU2 version 5 that has no viscous axisymmetric source term.
Now I use latest version where the issues are resolved.

In the issue #1106 , @jtneedels and @WallyMaier  ran verification cases using SA model below:
[](https://github.com/su2code/SU2/pull/1106#issuecomment-740223399)

But in my case, SA model in latest version show as like version 5 results (no axisymeetric source term).
When I see the source code, however, it seems that SST model axisymmetric source term exists, but SA model is not. 
`SU2_CFD/src/numerics/turbulent/turb_sources.cpp`

Can't I run axisymmetric case using SA model? 
I'm confused this. I really appreciate it if someone help me.

Thanks.

**Comment by WallyMaier** [comment]
Hi, you should be able to run the Axisymmetric and SA.  What type of case are you running/can you provide your cfg file?

  I believe the SST model requires more source terms (seen in turb_sources) to fully capture the relaxation effects.  @FlorianDm introduced the SST source terms, so he may be a better resource on this case.

**Reply by bigfooted** [reply]
@WallyMaier : In develop, I only see an implementation of ResidualAxisymmetric for SST, like @mskim0401 mentioned. Are the axisymmetry source terms for SA implemented in another branch?

**Reply by WallyMaier** [reply]
@bigfooted My understanding was the actual SA solver didnt require additional terms to determine the eddy viscosity.   @jtneedels and I only tested the laminar implementation.   So this assumption im making about the SA terms could be wrong.

**Reply by bigfooted** [reply]
Isn't SA implemented in conservation form as  (rho*nu)_t +  (rho*U*nu)_x + (rho*V*nu)_y = ...?
Then the gradient of y will give a source term  rho*V*nu/y 
And doesn't the diffusion term also give a contribution? I don't see why it is different than the SST model, or am I missing something?

**Reply by bigfooted** [reply]
I'm probably missing something....
In the SST model we add (among other terms) a convective term to the residual:  -yinv.Volume.rhov.k
For the temperature equation for instance, a similar term is +yinv.Volume.rho.V.cp.T
Why the change in sign? Don't these terms both come from the convective term (1/r).(rho.r.V.phi)_r?
The incompressible axisymmetric code gives the correct results for a laminar pipe flow, so at least the signs of those source terms are correct.

**Reply by WallyMaier** [reply]
Thats a good question.  I only looked into the contribution of eddy_viscosity to the axisymmetric source terms bulk fluid solver.  I havent looked into the actual turbulence solvers.

**Reply by mskim0401** [reply]
I'm sorry for late reply due to time difference and thank you for reply!
I'm trying to run axisymmetric base flow case with steady compressible RANS using SA model.

my grid is:
![스크린샷, 2021-06-29 10-30-40](https://user-images.githubusercontent.com/74228020/123723581-a139aa80-d8c5-11eb-8b29-9dc0fb3eb9f7.png)

I attached my configure file.
[mskim_configure.zip](https://github.com/su2code/SU2/files/6729923/mskim_configure.zip)

**Reply by WallyMaier** [reply]
@mskim0401 your config file looks good. You mentioned that the results you are getting matching inviscid results, or did I misunderstand?

**Reply by mskim0401** [reply]
Oh, I'm sorry. The results between SU2 version 5 (w/o viscous source term) and version 7 (with viscous source term) are not exactly matched, but far from experiment and other CFD result. 
The results are nonphysical. I judged by looking positive Base pressure coefficient.

![Base_Cp](https://user-images.githubusercontent.com/74228020/123728540-540e0680-d8ce-11eb-9f7a-c795895e67fc.png)
Both are axisymmetric cases.

SST result in version 7 is also different from other CFD result, but is acceptable.
Thank you.

**Reply by FlorianDm** [reply]
Hi everyone, yes I implemented the additional axisymmetry source terms for SST here https://github.com/su2code/SU2/pull/1195 where I added relevant literature which should allow to verify the correctness easily. I did not work on SA.

@bigfooted As stated in that PR: "Note that these source terms have the opposite sign (RHS) of the flow equations source terms (LHS) since the block is subtracted and not added." so this should answer your question if I understand correctly :).

@WallyMaier What do you mean by "the SST model requires more source terms (seen in turb_sources) to fully capture the relaxation effects"? Are you referring to the axisymmetry source terms only or the normal source terms? To me it seems all consistent with [menter1994.pdf](https://github.com/su2code/SU2/files/5957873/menter1994.pdf).

@mskim0401 In my own (limited) experience the SST axisymmetry source terms only have a minor effect on the solution while the flow equation axisymmetry source terms are absolutely essential. One thing I did notice in general was a high sensitivity of dissipation to the BL refinement (really needed to go down to y+ near 1) but I guess that should not be much of a surprise. I compared only an indicator of overall dissipation to a commercial CFD code which was almost identical.

**Reply by bigfooted** [reply]
@FlorianDm Ok, I still think the sign is wrong in the axisymmetry source terms then, maybe because of some -(-Q) source term implementation confusion. Please correct me if I'm wrong. This is how I think it works: 

If we look at the paper of Economon, we see in eq. (1) a source term -Q on the left hand side of the equation:

https://su2code.github.io/documents/AIAA-2018-3111.pdf

I think this minus sign is a bit misleading. We actually move all the terms (convection,diffusion, source terms) to the left hand side of the equation and the source term is *added* to the residual in the code, not subtracted, so we have convection+diffusion+sourceterm = 0 (not sure about the diffusion term, I think that one is subtracted?).
The axisymmetry source terms in eq. (21) in the paper all start with the negative convective source term. Together with the negative sign of Q in eq. (1), the convective source terms contributes as a positive term to the residual. We see that in flow_sources.cpp these axisymmetric source terms are positive. So the source term is on the left side of the equation with a positive sign. You should move all source terms in your reference equation for k and omega to the left hand side, and these are the signs to use. Now you have implemented a minus sign: cdk_axi = rhov.k - ... and residual+=yinv.Volume.(pk_axi - cdk_axi) 

I hope this is clear? And yes, it is probably a small effect but we should make sure that it is correct (and maybe add an axisymmetric turbulent case).

**Reply by WallyMaier** [reply]
@FlorianDm, @all.  Please disregard my previous posts, I was mistaken.

**Reply by FlorianDm** [reply]
@bigfooted This is indeed confusing so I checked it a bunch of times when I did it and now again and I really think it is correct. The signs are the same as those of the normal turbulent sources when you compare the production terms: 
https://github.com/su2code/SU2/blob/1f921473ea5a813e79f8f28b4bf6b1593bda34d5/SU2_CFD/src/numerics/turbulent/turb_sources.cpp#L884-L885  https://github.com/su2code/SU2/blob/1f921473ea5a813e79f8f28b4bf6b1593bda34d5/SU2_CFD/include/numerics/turbulent/turb_sources.hpp#L365-L366

For the (averaged) flow equations the sources block is **added** (LHS):
https://github.com/su2code/SU2/blob/1f921473ea5a813e79f8f28b4bf6b1593bda34d5/SU2_CFD/src/solvers/CEulerSolver.cpp#L2406

For the turbulence equations the sources block is **subtracted** (RHS):
https://github.com/su2code/SU2/blob/1f921473ea5a813e79f8f28b4bf6b1593bda34d5/SU2_CFD/src/solvers/CTurbSSTSolver.cpp#L347

Or I am still missing something...

**Reply by bigfooted** [reply]
OK, now I see where the difference comes from. But then the question is why we add source terms for the flow and subtract source terms for turbulence. This is really confusing...

**Reply by bigfooted** [reply]
Ok, so I am now convinced that the source terms for SST are correct. I think we can add the Jacobian as well. We take the gradient of the residual wrt rho.k and rho.w and then J[0][0]= (-v + P/rhok).Vol/r and J[1][1]=-v.Vol/r (ignoring derivatives of the gradients).

And then for SA, we need to add at least the convection-diffusion terms, and some other terms perhaps depending on the production terms.

**Reply by FlorianDm** [reply]
I had actually added the Jacobian in the PR but ended up removing it after testing a lot and not finding any (positive) effect on convergence. There is a little discussion with Pedro in the PR about conserving diagonal dominance but in the end it did not make a difference. You can find the removed Jacobian terms here https://github.com/su2code/SU2/pull/1195/commits/dd1b1b39e0ea33a85dc06d8e415d2e242d4ce410
